Fix #195 --no-fallback by default

anthonydahanne opened the pull request:
## Summary
* Remove deprecated arguments BP_BOOT_NATIVE_IMAGE_BUILD_ARGUMENTS and BP_BOOT_NATIVE_IMAGE
They've been deprecated for 2 years now
* Fix #195: use --no-fallback by default
* unless user has explicitly specified --auto-fallback or --force-fallback

## Usage:

```
> pack build -p ~/workspaces/paketo-buildpacks/samples/java/native-image/java-native-image-sample my-native   --env BP_NATIVE_IMAGE=true  --env BP_MAVEN_BUILD_ARGUMENTS="-Dmaven.test.skip=true --no-transfer-progress package -Pnative" -b paketo-buildpacks/syft \
-b paketo-buildpacks/bellsoft-liberica -b paketo-buildpacks/maven \
-b paketo-buildpacks/executable-jar -b . \
 --builder paketobuildpacks/builder:base --verbose --clear-cache
[...]
  Native Image: Contributing to layer
    Executing native-image --no-fallback -H:Name=/layers/paketo-buildpacks_native-image/native-image/io.paketo.demo.DemoApplication -cp /workspace io.paketo.demo.DemoApplication
[...]
```

```
pack build -p ~/workspaces/paketo-buildpacks/samples/java/native-image/java-native-image-sample plop   --env BP_
NATIVE_IMAGE=true  --env BP_MAVEN_BUILD_ARGUMENTS="-Dmaven.test.skip=true --no-transfer-progress package -Pnative" --env BP_NATIVE_IMAGE_BUILD_ARGUMENTS="--auto-fallback"  -b paketo-buildpacks/syft \ 
-b paketo-buildpacks/bellsoft-liberica -b paketo-buildpacks/maven \
-b paketo-buildpacks/executable-jar -b . \
 --builder paketobuildpacks/builder:base --verbose --clear-cache
[...]
  Native Image: Contributing to layer
    Executing native-image --auto-fallback -H:Name=/layers/paketo-buildpacks_native-image/native-image/io.paketo.demo.DemoApplication -cp /workspace io.paketo.demo.DemoApplication
[...]

```
## Checklist
<!-- Please confirm the following -->
* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [x] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

anthonydahanne commented:
@dmikusa hey there 👋 
I'm wondering about 2 things:
* I've imported a `golang.org/x/exp` - if you're not comfortable with this, let me know, I'll refactor
* I've just set minor version upgrade; even though I have removed some deprecated flags - let me know if I have to remove the deprecation removal commit of if it should be major bump

dmikusa commented:
> I've just set minor version upgrade; even though I have removed some deprecated flags - let me know if I have to remove the deprecation removal commit of if it should be major bump

I think we've removed settings before after deprecating them for a long period of time in minor versions. If you could separate this into its own PR, then we can wait till @pivotal-david-osullivan is back and double-check with him.

> I've imported a golang.org/x/exp - if you're not comfortable with this, let me know, I'll refactor

`slices.Contains` 🤯  I can't understand why this isn't in the go stdlib 

I'm pretty sure there are at least a couple of buildpacks or modules that have implemented "Contains". I have always intended to put that into libpak, but haven't had the time. That would be my preference, if we added a slices module we could put a common Contains in there. I can understand if you would rather just implement contains here. I'm fine with that as well.

I'm not necessarily adverse to a 3rd party library if there is perhaps another one, but what I don't like about this particular one is a.) that it's listed as experimental b.) it's kind of a grab bag with all sorts of functionality some totally unrelated, I'd prefer something that's specific to just slices or at least collections c.) there's no 1.0 release.

anthonydahanne commented:
Hello @dmikusa !
Thanks for your comments!

* I have moved the deprecation removal commit to another PR #256 
* I have added a new `slices` package to this PR (that copies the 2 needed functions for x/exp); meanwhile I've created an issue in libpak to create such `slices` package in it, https://github.com/paketo-buildpacks/libpak/issues/238

dmikusa commented on native/slices/slices.go:
```diff
@@ -0,0 +1,20 @@
+package slices
+
+// Those 2 functions were copied from https://cs.opensource.google/go/x/exp/+/master:slices/slices.go ,
+// an experimental collection of utility methods that could be promoted to stdlib in the future
```
```suggestion
// Those 2 functions were copied from https://cs.opensource.google/go/x/exp/+/master:slices/slices.go ,
// an experimental collection of utility methods that could be promoted to stdlib in the future

Copyright (c) 2009 The Go Authors. All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

   * Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.
   * Redistributions in binary form must reproduce the above
copyright notice, this list of conditions and the following disclaimer
in the documentation and/or other materials provided with the
distribution.
   * Neither the name of Google Inc. nor the names of its
contributors may be used to endorse or promote products derived from
this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```

Not a lawyer, but I think we need to include the license to be compliant.

anthonydahanne replied:
yeah, probably - ok!

dmikusa commented on native/native_image.go:
```diff
@@ -69,9 +70,14 @@ func (n NativeImage) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
 		return libcnb.Layer{}, fmt.Errorf("unable to process arguments\n%w", err)
 	}
+
+	if !slices.Contains(arguments, "--auto-fallback") && !slices.Contains(arguments, "--force-fallback") {
```
Is there a test case that includes `--auto-fallback` or `--force-fallback`?  I don't see it in the commit log.

anthonydahanne commented:
Yes, I forgot some tests; should be good now! please let me know!
